### PR TITLE
Make gettext functions filterable

### LINKED ISSUE #12516: Allow filtering internationalized strings
In WordPress core, every single string that is passed to `__()` & co. can be filtered using the `gettext` filter in `translate()`.

Arguments to that filter are the original text, the translation, and the text domain.

For ultimate flexibility, I suggest adding the same filters to gettext functions in the i18n package.

### PR DESCRIPTION
**Pull request by swissspidy**
## Description
Fixes #12516 by applying filters to all the gettext function in the i18n package.

## How has this been tested?
Manually added some filters to these hooks.

## Types of changes
New feature (non-breaking change which adds functionality): added new filters.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by aduth** [comment]
Personally, I don't find this to be a pressing enhancement to consider, and while normally I'd not have such reservations, the translate functions are very hot paths. `applyFilters` is not especially "expensive", but without a use-case beyond broad flexibility and feature parity, it doesn't feel very compelling.

Alternatively, we could consider optimizing for the assumption that a filter will not be present in the majority of usage, and use a monkey-patching technique in overriding the translation functions when a filter is applied (see https://github.com/WordPress/gutenberg/issues/8602#issuecomment-432227250).

**Comment by swissspidy** [comment]
Yeah it's definitely not a pressing enhancement, but sure something that will come up when JS I18N is getting more traction now with 5.0 and beyond. Parity and flexibility for devs is definitely important to some degree, IMO.

Your monkey-patching idea sounds interesting, but I guess we would need to check first whether it would be worth the effort. I was thinking about doing some benchmarks or something, but I wouldn't know where to start there.

**Comment by swissspidy** [comment]
For the record, we discussed some use cases at WCUS, mainly #12407 and https://core.trac.wordpress.org/ticket/45425#comment:10. It's certainly not ideal that the user experience for certain locales got worse.

As for the technique used here, we could do some benchmarks to see whether monkey-patching should be explored or not. Although that could also be done in a later version.

File: packages/i18n/src/index.js [hunk 1]
```diff
@@ -90,7 +95,16 @@ function dcnpgettext( domain = 'default', context, single, plural, number ) {
...
+	const translation = dcnpgettext( domain, undefined, text );
+
+	/**
+	 * Filters text with its translation.
+	 *
+	 * @param {string} translation Translated text.
+	 * @param {string} text        Text to translate.
+	 * @param {string} domain      Text domain. Unique identifier for retrieving translated strings.
+	 */
+	return applyFilters( 'i18n.gettext', translation, text, domain );
```
Comment (gziolo) [inline packages/i18n/src/index.js]: @aduth - we probably should also optimize `applyFilters` to return early when there are no hooks to process. See:

https://github.com/WordPress/gutenberg/blob/master/packages/hooks/src/createRunHook.js#L22-L33

It makes me think that having to run filters on every re-render might be bad for the performance of the application.

Comment (gziolo) [reply packages/i18n/src/index.js]: Okey, I see https://github.com/WordPress/gutenberg/pull/12517#issuecomment-444561526 which aligns with my comment.

Comment (aduth) [reply packages/i18n/src/index.js]: Yeah, it's not too horrible though. Outside of some conditions and a one-time setup, it amounts to incrementing the `runs` each time it's called.

**Comment by gziolo** [comment]
It looks like this is something that we need to decide about whether we continue working on this PR or we close it and keep it as a future reference when we decide it becomes necessary.

**Comment by swissspidy** [comment]
@gziolo @aduth What's the way forward here? A discussion at the next weekly meeting together with some core i18n / polyglots people involved?

**Comment by aduth** [comment]
> A discussion at the next weekly meeting together with some core i18n / polyglots people involved?

Which weekly meeting are you referring to? I can have it included it in the agenda of tomorrow's JavaScript chat if you'd like.

**Comment by swissspidy** [comment]
Yeah that could work.

**Comment by aduth** [comment]
Recap from aforementioned meeting: https://make.wordpress.org/core/2019/04/10/javascript-chat-summary-april-9th-2019/

**Pull request closed without merging**